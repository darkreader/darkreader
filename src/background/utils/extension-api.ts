import {isPDF} from '../../utils/url';
import {isFirefox, isEdge} from '../../utils/platform';

declare const browser: {
    commands: {
        update({name, shortcut}): void;
    };
};

export function canInjectScript(url: string) {
    if (isFirefox) {
        return (url
            && !url.startsWith('about:')
            && !url.startsWith('moz')
            && !url.startsWith('view-source:')
            && !url.startsWith('https://addons.mozilla.org/')
            && !isPDF(url)
        );
    }
    if (isEdge) {
        return (url
            && !url.startsWith('chrome')
            && !url.startsWith('data')
            && !url.startsWith('devtools')
            && !url.startsWith('edge')
            && !url.startsWith('https://chrome.google.com/webstore')
            && !url.startsWith('https://microsoftedge.microsoft.com/addons')
            && !url.startsWith('view-source')
        );
    }
    return (url
        && !url.startsWith('chrome')
        && !url.startsWith('https://chrome.google.com/webstore')
        && !url.startsWith('data')
        && !url.startsWith('devtools')
        && !url.startsWith('view-source')
    );
}

let isWriting = false;

export async function readSyncStorage<T extends {[key: string]: any}>(defaults: T): Promise<T> {
    return new Promise<T>((resolve) => {
        chrome.storage.sync.get(defaults, (sync: T) => {
            if (chrome.runtime.lastError) {
                console.error(chrome.runtime.lastError.message);
                resolve(defaults);
                return;
            }
            resolve(sync);
        });
    });
}

export async function readLocalStorage<T extends {[key: string]: any}>(defaults: T): Promise<T> {
    return new Promise<T>((resolve) => {
        chrome.storage.local.get(defaults, (local: T) => {
            if (chrome.runtime.lastError) {
                console.error(chrome.runtime.lastError.message);
                resolve(defaults);
                return;
            }
            resolve(local);
        });
    });
}

export async function writeSyncStorage<T extends {[key: string]: any}>(values: T): Promise<void> {
    return new Promise<void>((resolve, reject) => {
        isWriting = true;
        chrome.storage.sync.set(values, () => {
            if (chrome.runtime.lastError) {
                reject(chrome.runtime.lastError);
                return;
            }
            resolve();
            setTimeout(() => isWriting = false);
        });
    });
}

export async function writeLocalStorage<T extends {[key: string]: any}>(values: T): Promise<void> {
    return new Promise<void>((resolve) => {
        isWriting = true;
        chrome.storage.local.set(values, () => {
            resolve();
            setTimeout(() => isWriting = false);
        });
    });
}

export const subscribeToOuterSettingsChange = (callback: () => void) => {
    chrome.storage.onChanged.addListener(() => {
        if (!isWriting) {
            callback();
        }
    });
};

export async function getFontList() {
    return new Promise<string[]>((resolve) => {
        if (!chrome.fontSettings) {
            // Todo: Remove it as soon as Firefox and Edge get support.
            resolve([
                'serif',
                'sans-serif',
                'monospace',
                'cursive',
                'fantasy',
                'system-ui'
            ]);
            return;
        }
        chrome.fontSettings.getFontList((list) => {
            const fonts = list.map((f) => f.fontId);
            resolve(fonts);
        });
    });
}

export async function getCommands() {
    return new Promise<chrome.commands.Command[]>((resolve) => {
        if (!chrome.commands) {
            resolve([]);
            return;
        }
        chrome.commands.getAll((commands) => {
            if (commands) {
                resolve(commands);
            } else {
                resolve([]);
            }
        });
    });
}

export function setShortcut(command: string, shortcut: string) {
    if (typeof browser !== 'undefined' && browser.commands && browser.commands.update) {
        browser.commands.update({name: command, shortcut});
    }
}
