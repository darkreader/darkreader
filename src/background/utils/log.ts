import {sendLog} from './sendLog';

declare const __DEBUG__: boolean;
declare const __TEST__: boolean;

export function logInfo(...args: any[]): void {
    if (__DEBUG__) {
        console.info(...args);
        sendLog('info', args);
    }
}

export function logWarn(...args: any[]): void {
    if (__DEBUG__) {
        console.warn(...args);
        sendLog('warn', args);
    }
}

export function logInfoCollapsed(title: string, ...args: any[]): void {
    if (__DEBUG__) {
        console.groupCollapsed(title);
        console.log(...args);
        console.groupEnd();
        sendLog('info', args);
    }
}

function logAssert(...args: any[]): void {
    if ((__TEST__ || __DEBUG__)) {
        console.assert(...args);
        sendLog('assert', ...args);
    }
}

export function ASSERT(description: string, condition: (() => boolean) | any): void {
    if ((__TEST__ || __DEBUG__) && (typeof condition === 'function' && !condition()) || !Boolean(condition)) {
        logAssert(description);
        if (__TEST__) {
            throw new Error(`Assertion failed: ${description}`);
        }
    }
}
