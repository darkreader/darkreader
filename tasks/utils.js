const colors = Object.entries({
    gray: '\x1b[90m',
    green: '\x1b[32m',
    red: '\x1b[31m',
    yellow: '\x1b[33m',
}).reduce((map, [key, value]) => Object.assign(map, {[key]: (text) => `${value}${text}\x1b[0m`}), {});

function logWithTime(text) {
    const now = new Date();
    const hours = now.getHours();
    const minutes = now.getMinutes();
    const seconds = now.getSeconds();
    const leftpad = (n) => String(n).padStart(2, '0');
    return console.log(`${colors.gray([hours, minutes, seconds].map(leftpad).join(':'))} ${text}`);
}

const log = Object.assign((text) => logWithTime(text), {
    ok: (text) => logWithTime(colors.green(text)),
    warn: (text) => logWithTime(colors.yellow(text)),
    error: (text) => logWithTime(colors.red(text)),
});

module.exports = {
    log,
};
