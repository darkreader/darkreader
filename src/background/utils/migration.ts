import {isChromium} from '../../utils/platform';

export function isNonPersistent() {
    if (!isChromium) {
        return false;
    }
    const background = chrome.runtime.getManifest().background;
    if ('persistent' in background) {
        return background.persistent === false;
    }
    if ('service_worker' in background) {
        return true;
    }
}
