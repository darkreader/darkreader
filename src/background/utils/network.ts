import {loadAsDataURL, loadAsText} from '../../utils/network';
import {getStringSize} from '../../utils/text';
import {getDuration} from '../../utils/time';

interface RequestParams {
    url: string;
    timeout?: number;
}

export async function readText(params: RequestParams): Promise<string> {
    return new Promise((resolve, reject) => {
        const request = new XMLHttpRequest();
        request.overrideMimeType('text/plain');
        request.open('GET', params.url, true);
        request.onload = () => {
            if (request.status >= 200 && request.status < 300) {
                resolve(request.responseText);
            } else {
                reject(new Error(`${request.status}: ${request.statusText}`));
            }
        };
        request.onerror = () => reject(new Error(`${request.status}: ${request.statusText}`));
        if (params.timeout) {
            request.timeout = params.timeout;
            request.ontimeout = () => reject(new Error('File loading stopped due to timeout'));
        }
        request.send();
    });
}

interface CacheRecord {
    expires: number;
    size: number;
    url: string;
    value: string;
}

class LimitedCacheStorage {
    static QUOTA_BYTES = ((navigator as any).deviceMemory || 4) * 16 * 1024 * 1024;
    static TTL = getDuration({minutes: 10});
    static ALARM_NAME = 'network';

    private bytesInUse = 0;
    private records = new Map<string, CacheRecord>();

    constructor() {
        chrome.alarms.onAlarm.addListener(async (alarm) => {
            if (alarm.name === LimitedCacheStorage.ALARM_NAME) {
                this.removeExpiredRecords();
            }
        });
        chrome.alarms.create(LimitedCacheStorage.ALARM_NAME, {periodInMinutes: 1});
    }

    has(url: string) {
        return this.records.has(url);
    }

    get(url: string) {
        if (this.records.has(url)) {
            const record = this.records.get(url);
            record.expires = Date.now() + LimitedCacheStorage.TTL;
            this.records.delete(url);
            this.records.set(url, record);
            return record.value;
        }
        return null;
    }

    set(url: string, value: string) {
        const size = getStringSize(value);
        if (size > LimitedCacheStorage.QUOTA_BYTES) {
            return;
        }

        for (const [url, record] of this.records) {
            if (this.bytesInUse + size > LimitedCacheStorage.QUOTA_BYTES) {
                this.records.delete(url);
                this.bytesInUse -= record.size;
            } else {
                break;
            }
        }

        const expires = Date.now() + LimitedCacheStorage.TTL;
        this.records.set(url, {url, value, size, expires});
        this.bytesInUse += size;
    }

    private removeExpiredRecords() {
        const now = Date.now();
        for (const [url, record] of this.records) {
            if (record.expires < now) {
                this.records.delete(url);
                this.bytesInUse -= record.size;
            } else {
                break;
            }
        }
    }
}

interface FetchRequestParameters {
    url: string;
    responseType: 'data-url' | 'text';
    mimeType?: string;
    origin?: string;
}

export function createFileLoader() {
    const caches = {
        'data-url': new LimitedCacheStorage(),
        'text': new LimitedCacheStorage(),
    };

    const loaders = {
        'data-url': loadAsDataURL,
        'text': loadAsText,
    };

    async function get({url, responseType, mimeType, origin}: FetchRequestParameters) {
        const cache = caches[responseType];
        const load = loaders[responseType];
        if (cache.has(url)) {
            return cache.get(url);
        }

        const data = await load(url, mimeType, origin);
        cache.set(url, data);
        return data;
    }

    return {get};
}
