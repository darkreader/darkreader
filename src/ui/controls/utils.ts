import {classes} from '../utils';

function toArray<T>(x: T | T[]) {
    return Array.isArray(x) ? x : [x];
}

export function mergeClass(
    cls: string | {[cls: string]: any} | Array<string | {[cls: string]: any}>,
    propsCls: string | {[cls: string]: any} | Array<string | {[cls: string]: any}>
) {
    const normalized = toArray(cls).concat(toArray(propsCls));
    return classes(...normalized);
}

export function omitAttrs(omit: string[], attrs: Malevic.NodeAttrs) {
    const result: Malevic.NodeAttrs = {};
    Object.keys(attrs).forEach((key) => {
        if (omit.indexOf(key) < 0) {
            result[key] = attrs[key];
        }
    });
    return result;
}

export function isElementHidden(element: HTMLElement) {
    return element.offsetParent === null;
}
