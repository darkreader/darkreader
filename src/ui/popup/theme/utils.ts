import type {Theme} from '../../../definitions';
import {isURLInList} from '../../../utils/url';
import type {ViewProps} from '../types';

export function getCurrentThemePreset(props: ViewProps) {
    const custom = props.data.settings.customThemes.find(
        ({url}) => isURLInList(props.tab.url, url)
    );
    const preset = custom ? null : props.data.settings.presets.find(
        ({urls}) => isURLInList(props.tab.url, urls)
    );
    const theme = custom ?
        custom.theme :
        preset ?
            preset.theme :
            props.data.settings.theme;

    function setTheme(config: Partial<Theme>) {
        if (custom) {
            custom.theme = {...custom.theme, ...config};
            props.actions.changeSettings({
                customThemes: props.data.settings.customThemes,
            });
        } else if (preset) {
            preset.theme = {...preset.theme, ...config};
            props.actions.changeSettings({
                presets: props.data.settings.presets,
            });
        } else {
            props.actions.setTheme(config);
        }
    }

    return {
        theme,
        change: setTheme,
    };
}
