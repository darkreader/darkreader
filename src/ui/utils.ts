import {isMobile} from '../utils/platform';

declare const __CHROMIUM_MV3__: boolean;

export function classes(...args: Array<string | {[cls: string]: boolean}>): string {
    const classes: string[] = [];
    args.filter((c) => Boolean(c)).forEach((c) => {
        if (typeof c === 'string') {
            classes.push(c);
        } else if (typeof c === 'object') {
            classes.push(...Object.keys(c).filter((key) => Boolean(c[key])));
        }
    });
    return classes.join(' ');
}

export function compose<T extends Malevic.Component>(type: T, ...wrappers: Array<(t: T) => T>): T {
    return wrappers.reduce((t, w) => w(t), type);
}

export function openFile(options: {extensions: string[]}, callback: (content: string) => void): void {
    const input = document.createElement('input');
    input.type = 'file';
    input.style.display = 'none';
    if (options.extensions && options.extensions.length > 0) {
        input.accept = options.extensions.map((ext) => `.${ext}`).join(',');
    }
    const reader = new FileReader();
    reader.onloadend = () => callback(reader.result as string);
    input.onchange = () => {
        if (input.files![0]) {
            reader.readAsText(input.files![0]);
            document.body.removeChild(input);
        }
    };
    document.body.appendChild(input);
    input.click();
}

export function saveFile(name: string, content: string): void {
    const a = document.createElement('a');
    a.href = URL.createObjectURL(new Blob([content]));
    a.download = name;
    a.click();
}

type AnyVoidFunction = (...args: any[]) => void;

export function throttle<F extends AnyVoidFunction>(callback: F): F {
    let frameId: number | null = null;
    return ((...args: any[]) => {
        if (!frameId) {
            callback(...args);
            frameId = requestAnimationFrame(() => (frameId = null));
        }
    }) as F;
}

interface SwipeEventObject {
    clientX: number;
    clientY: number;
}

type SwipeEventHandler<T = void> = (e: SwipeEventObject, nativeEvent: MouseEvent | TouchEvent) => T;
type StartSwipeHandler = SwipeEventHandler<{move: SwipeEventHandler; up: SwipeEventHandler}>;

function onSwipeStart(
    startEventObj: MouseEvent | TouchEvent,
    startHandler: StartSwipeHandler,
): void {
    const isTouchEvent =
        typeof TouchEvent !== 'undefined' &&
        startEventObj instanceof TouchEvent;
    const touchId = isTouchEvent
        ? (startEventObj).changedTouches[0].identifier
        : null;
    const pointerMoveEvent = isTouchEvent ? 'touchmove' : 'mousemove';
    const pointerUpEvent = isTouchEvent ? 'touchend' : 'mouseup';

    if (!isTouchEvent) {
        startEventObj.preventDefault();
    }

    function getSwipeEventObject(e: MouseEvent | TouchEvent) {
        const {clientX, clientY} = isTouchEvent
            ? getTouch(e as TouchEvent)
            : e as MouseEvent;
        return {clientX, clientY};
    }

    const startSE = getSwipeEventObject(startEventObj);
    const {move: moveHandler, up: upHandler} = startHandler(startSE, startEventObj);

    function getTouch(e: TouchEvent) {
        return Array.from(e.changedTouches).find(
            ({identifier: id}) => id === touchId,
        )!;
    }

    const onPointerMove = throttle((e) => {
        const se = getSwipeEventObject(e);
        moveHandler(se, e);
    });

    function onPointerUp(e: MouseEvent) {
        unsubscribe();
        const se = getSwipeEventObject(e);
        upHandler(se, e);
    }

    function unsubscribe() {
        window.removeEventListener(pointerMoveEvent, onPointerMove);
        window.removeEventListener(pointerUpEvent, onPointerUp);
    }

    window.addEventListener(pointerMoveEvent, onPointerMove, {passive: true});
    window.addEventListener(pointerUpEvent, onPointerUp, {passive: true});
}

export function createSwipeHandler(startHandler: StartSwipeHandler): (e: MouseEvent | TouchEvent) => void {
    return (e: MouseEvent | TouchEvent) => onSwipeStart(e, startHandler);
}

export async function getFontList(): Promise<string[]> {
    return new Promise<string[]>((resolve) => {
        if (!chrome.fontSettings) {
            // Todo: Remove it as soon as Firefox and Edge get support.
            resolve([
                'serif',
                'sans-serif',
                'monospace',
                'cursive',
                'fantasy',
                'system-ui',
            ]);
            return;
        }
        chrome.fontSettings.getFontList((list) => {
            const fonts = list.map((f) => f.fontId);
            resolve(fonts);
        });
    });
}

type page = 'devtools' | 'stylesheet-editor';

// TODO(Anton): There must be a better way to do this
// This function ping-pongs a message to possible DevTools popups.
// This function should have reasonable performance since it sends
// messages only to popups and not regular windows.
async function getExtensionPageTabMV3(): Promise<chrome.tabs.Tab | null> {
    return new Promise((resolve) => {
        chrome.windows.getAll({
            populate: true,
            windowTypes: ['popup'],
        }, (w) => {
            const responses: Array<Promise<string>> = [];
            let found = false;
            for (const window of w) {
                const response = chrome.tabs.sendMessage<string, 'getExtensionPageTabMV3_pong'>(window.tabs![0]!.id!, 'getExtensionPageTabMV3_ping', {frameId: 0});
                response.then((response) => {
                    if (response === 'getExtensionPageTabMV3_pong') {
                        found = true;
                        resolve(window.tabs![0]);
                    }
                });
                responses.push(response);
            }
            Promise.all(responses).then(() => !found && resolve(null));
        });
    });
}

async function getExtensionPageTab(url: string): Promise<chrome.tabs.Tab | null> {
    if (__CHROMIUM_MV3__) {
        return getExtensionPageTabMV3();
    }
    return new Promise<chrome.tabs.Tab>((resolve) => {
        chrome.tabs.query({
            url,
        }, ([tab]) => resolve(tab || null));
    });
}

export async function openExtensionPage(page: page): Promise<void> {
    const url = chrome.runtime.getURL(`/ui/${page}/index.html`);
    if (isMobile) {
        const extensionPageTab = await getExtensionPageTab(url);
        if (extensionPageTab !== null) {
            chrome.tabs.update(extensionPageTab.id!, {active: true});
            window.close();
        } else {
            chrome.tabs.create({url});
            window.close();
        }
    } else {
        const extensionPageTab = await getExtensionPageTab(url);
        if (extensionPageTab !== null) {
            chrome.windows.update(extensionPageTab.windowId, {focused: true});
            window.close();
        } else {
            chrome.windows.create({
                type: 'popup',
                url,
                width: 600,
                height: 600,
            });
            window.close();
        }
    }
}
