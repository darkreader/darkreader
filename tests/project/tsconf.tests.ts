import ts, {ExitStatus} from 'typescript';
import {dirname, join} from 'path';
import {fork} from 'child_process';
import {tmpdir} from 'os';
import {mkdtemp} from 'fs/promises';
import {childClosed} from './utils';

const rootDir = dirname(require.resolve('../../package.json'));
const tsc = require.resolve('typescript/bin/tsc');

const tsProjects: Array<{tsconfig: string}> = [
    // Browser extension
    {tsconfig: 'src'},
    // API
    {tsconfig: 'src/api'},
    // E2E tests (Jest, Puppeteer, Node.js)
    // TODO(anton): fix upstream type conflict and uncomment me
    // {tsconfig: 'tests/browser'},
    // Browser tests (Karma, Jasmine)
    {tsconfig: 'tests/inject'},
    // Unit tests (Jest, Node.js)
    {tsconfig: 'tests/unit'},
];

describe('TypeScript project config', () => {
    it.each(tsProjects)('file should parse and resolve correctly: $tsconfig', async ({tsconfig}) => {
        // Parse config from temp dir instead of root dir
        const cwd = await mkdtemp(join(tmpdir(), 'darkreader'));
        const project = join(rootDir, tsconfig, 'tsconfig.json');

        // Fire
        const child = fork(tsc, ['--project', project, '--showConfig'], {silent: true, cwd});

        const {response} = await childClosed(child, {serialization: 'json'});
        expect(child.exitCode).toBe(ExitStatus.Success);
        const config = ts.parseJsonConfigFileContent(response, ts.sys, project);
        expect(config.raw).toMatchSnapshot();
    }, 100000);

    // Slow test (run using `npm run test:project`)
    it.each(tsProjects)('should compile without errors: $tsconfig', async ({tsconfig}) => {
        // Compile config from temp dir instead of root dir
        const cwd = await mkdtemp(join(tmpdir(), 'darkreader'));
        const project = join(rootDir, tsconfig, 'tsconfig.json');

        // Fire
        const child = fork(tsc, ['--project', project, '--noEmit'], {silent: true, cwd});
        await childClosed(child);
        expect(child.exitCode).toBe(ExitStatus.Success);
    }, 60000);
});
