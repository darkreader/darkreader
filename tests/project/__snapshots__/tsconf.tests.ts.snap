// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TypeScript project config file should parse and resolve correctly: src 1`] = `
Object {
  "compileOnSave": false,
  "compilerOptions": Object {
    "allowJs": true,
    "baseUrl": "./",
    "downlevelIteration": true,
    "jsx": "react",
    "jsxFactory": "m",
    "lib": Array [
      "es6",
      "es2017",
      "dom",
      "dom.iterable",
    ],
    "module": "es6",
    "moduleResolution": "node",
    "noEmit": true,
    "noImplicitAny": true,
    "resolveJsonModule": true,
    "target": "es2019",
    "types": Array [
      "chrome",
      "offscreencanvas",
    ],
  },
  "files": Array [
    "./defaults.ts",
    "./definitions.d.ts",
    "./api/chrome.ts",
    "./api/fetch.ts",
    "./api/index.ts",
    "./background/config-manager.ts",
    "./background/devtools.ts",
    "./background/extension.ts",
    "./background/icon-manager.ts",
    "./background/index.ts",
    "./background/make-chromium-happy.ts",
    "./background/messenger.ts",
    "./background/newsmaker.ts",
    "./background/tab-manager.ts",
    "./background/user-storage.ts",
    "./background/window-theme.ts",
    "./background/utils/extension-api.ts",
    "./background/utils/migration.ts",
    "./background/utils/network.ts",
    "./background/utils/state-manager.ts",
    "./generators/css-filter.ts",
    "./generators/dynamic-theme.ts",
    "./generators/modify-colors.ts",
    "./generators/static-theme.ts",
    "./generators/svg-filter.ts",
    "./generators/text-style.ts",
    "./generators/theme-engines.ts",
    "./generators/utils/format.ts",
    "./generators/utils/matrix.ts",
    "./generators/utils/parse.ts",
    "./inject/detector.ts",
    "./inject/fallback.ts",
    "./inject/index.ts",
    "./inject/style.ts",
    "./inject/svg-filter.ts",
    "./inject/dynamic-theme/adopted-style-manger.ts",
    "./inject/dynamic-theme/css-collection.ts",
    "./inject/dynamic-theme/css-rules.ts",
    "./inject/dynamic-theme/image.ts",
    "./inject/dynamic-theme/index.ts",
    "./inject/dynamic-theme/inline-style.ts",
    "./inject/dynamic-theme/meta-theme-color.ts",
    "./inject/dynamic-theme/modify-css.ts",
    "./inject/dynamic-theme/network.ts",
    "./inject/dynamic-theme/style-manager.ts",
    "./inject/dynamic-theme/stylesheet-modifier.ts",
    "./inject/dynamic-theme/stylesheet-proxy.ts",
    "./inject/dynamic-theme/variables.ts",
    "./inject/dynamic-theme/watch.ts",
    "./inject/utils/dom.ts",
    "./inject/utils/watch-color-scheme.ts",
    "./ui/utils.ts",
    "./ui/connect/connector.ts",
    "./ui/connect/mock.ts",
    "./ui/controls/index.ts",
    "./ui/controls/utils.ts",
    "./ui/controls/button/index.tsx",
    "./ui/controls/checkbox/index.tsx",
    "./ui/controls/color-dropdown/index.tsx",
    "./ui/controls/color-picker/hsb-picker.tsx",
    "./ui/controls/color-picker/index.tsx",
    "./ui/controls/dropdown/index.tsx",
    "./ui/controls/message-box/index.tsx",
    "./ui/controls/multi-switch/index.tsx",
    "./ui/controls/nav-button/index.tsx",
    "./ui/controls/overlay/index.ts",
    "./ui/controls/reset-button/index.tsx",
    "./ui/controls/select/index.tsx",
    "./ui/controls/shortcut/index.tsx",
    "./ui/controls/slider/index.tsx",
    "./ui/controls/tab-panel/index.tsx",
    "./ui/controls/tab-panel/tab.tsx",
    "./ui/controls/text-list/index.tsx",
    "./ui/controls/textbox/index.tsx",
    "./ui/controls/time-range-picker/index.tsx",
    "./ui/controls/toggle/index.tsx",
    "./ui/controls/updown/index.tsx",
    "./ui/controls/updown/track.tsx",
    "./ui/controls/virtual-scroll/index.tsx",
    "./ui/devtools/index.tsx",
    "./ui/devtools/components/body.tsx",
    "./ui/popup/compatibility.js",
    "./ui/popup/index.tsx",
    "./ui/popup/types.d.ts",
    "./ui/popup/automation-page/index.tsx",
    "./ui/popup/body/index.tsx",
    "./ui/popup/check-button/index.tsx",
    "./ui/popup/components/body.tsx",
    "./ui/popup/components/custom-settings-toggle/index.tsx",
    "./ui/popup/components/engine-switch/index.tsx",
    "./ui/popup/components/filter-settings/index.tsx",
    "./ui/popup/components/filter-settings/mode-toggle.tsx",
    "./ui/popup/components/font-settings/index.tsx",
    "./ui/popup/components/header/index.tsx",
    "./ui/popup/components/header/more-toggle-settings.tsx",
    "./ui/popup/components/loader/index.tsx",
    "./ui/popup/components/more-settings/index.tsx",
    "./ui/popup/components/news/index.tsx",
    "./ui/popup/components/site-list-settings/index.tsx",
    "./ui/popup/components/site-toggle/checkmark-icon.tsx",
    "./ui/popup/components/site-toggle/index.tsx",
    "./ui/popup/control-group/index.tsx",
    "./ui/popup/main-page/app-switch.tsx",
    "./ui/popup/main-page/help.tsx",
    "./ui/popup/main-page/index.tsx",
    "./ui/popup/main-page/site-toggle.tsx",
    "./ui/popup/main-page/sun-moon-icon.tsx",
    "./ui/popup/main-page/system-icon.tsx",
    "./ui/popup/main-page/theme-group.tsx",
    "./ui/popup/main-page/watch-icon.tsx",
    "./ui/popup/manage-settings-page/export-settings.tsx",
    "./ui/popup/manage-settings-page/export-theme.tsx",
    "./ui/popup/manage-settings-page/fetch-news.tsx",
    "./ui/popup/manage-settings-page/import-settings.tsx",
    "./ui/popup/manage-settings-page/index.tsx",
    "./ui/popup/manage-settings-page/reset-settings-button.tsx",
    "./ui/popup/manage-settings-page/sync-config.tsx",
    "./ui/popup/manage-settings-page/sync-settings.tsx",
    "./ui/popup/news-section/index.tsx",
    "./ui/popup/page-viewer/index.tsx",
    "./ui/popup/settings-page/automation-button.tsx",
    "./ui/popup/settings-page/change-browser-theme.tsx",
    "./ui/popup/settings-page/context-menus.tsx",
    "./ui/popup/settings-page/detect-dark-theme.tsx",
    "./ui/popup/settings-page/devtools.tsx",
    "./ui/popup/settings-page/enabled-by-default.tsx",
    "./ui/popup/settings-page/index.tsx",
    "./ui/popup/settings-page/mange-settings-button.tsx",
    "./ui/popup/settings-page/site-list-button.tsx",
    "./ui/popup/settings-page/version.tsx",
    "./ui/popup/site-list-page/index.tsx",
    "./ui/popup/site-list-page/site-list.tsx",
    "./ui/popup/theme/utils.ts",
    "./ui/popup/theme/controls/background-color.tsx",
    "./ui/popup/theme/controls/brightness.tsx",
    "./ui/popup/theme/controls/color-scheme.tsx",
    "./ui/popup/theme/controls/contrast.tsx",
    "./ui/popup/theme/controls/font-picker.tsx",
    "./ui/popup/theme/controls/format.ts",
    "./ui/popup/theme/controls/grayscale.tsx",
    "./ui/popup/theme/controls/immediate-modify.tsx",
    "./ui/popup/theme/controls/index.tsx",
    "./ui/popup/theme/controls/mode.tsx",
    "./ui/popup/theme/controls/reset-button.tsx",
    "./ui/popup/theme/controls/scheme.tsx",
    "./ui/popup/theme/controls/scrollbar.tsx",
    "./ui/popup/theme/controls/selection.tsx",
    "./ui/popup/theme/controls/sepia.tsx",
    "./ui/popup/theme/controls/style-system-controls.tsx",
    "./ui/popup/theme/controls/text-color.tsx",
    "./ui/popup/theme/controls/text-stroke.tsx",
    "./ui/popup/theme/controls/theme-control.tsx",
    "./ui/popup/theme/controls/use-font.tsx",
    "./ui/popup/theme/page/collapsible-panel.tsx",
    "./ui/popup/theme/page/index.tsx",
    "./ui/popup/theme/preset-picker/index.tsx",
    "./ui/popup/utils/issues.ts",
    "./ui/popup/utils/markdown.tsx",
    "./ui/stylesheet-editor/index.tsx",
    "./ui/stylesheet-editor/components/body.tsx",
    "./utils/array.ts",
    "./utils/async-queue.ts",
    "./utils/color.ts",
    "./utils/colorscheme-parser.ts",
    "./utils/debounce.ts",
    "./utils/ipv6.ts",
    "./utils/links.ts",
    "./utils/locales.ts",
    "./utils/log.ts",
    "./utils/math.ts",
    "./utils/media-query.ts",
    "./utils/message.ts",
    "./utils/network.ts",
    "./utils/parsing.ts",
    "./utils/platform.ts",
    "./utils/promise-barrier.ts",
    "./utils/text.ts",
    "./utils/throttle.ts",
    "./utils/time.ts",
    "./utils/uid.ts",
    "./utils/url.ts",
  ],
}
`;

exports[`TypeScript project config file should parse and resolve correctly: src/api 1`] = `
Object {
  "compileOnSave": false,
  "compilerOptions": Object {
    "allowJs": true,
    "baseUrl": "..",
    "downlevelIteration": true,
    "jsx": "react",
    "jsxFactory": "m",
    "lib": Array [
      "es6",
      "es2017",
      "dom",
      "dom.iterable",
    ],
    "module": "es6",
    "moduleResolution": "node",
    "noEmit": true,
    "noImplicitAny": true,
    "resolveJsonModule": true,
    "target": "es5",
    "types": Array [
      "chrome",
      "offscreencanvas",
    ],
  },
  "files": Array [
    "./chrome.ts",
    "./fetch.ts",
    "./index.ts",
  ],
}
`;

exports[`TypeScript project config file should parse and resolve correctly: tests/browser 1`] = `
Object {
  "compileOnSave": false,
  "compilerOptions": Object {
    "allowJs": true,
    "downlevelIteration": true,
    "esModuleInterop": true,
    "jsx": "react",
    "jsxFactory": "m",
    "lib": Array [
      "es6",
      "dom",
    ],
    "module": "commonjs",
    "moduleResolution": "node",
    "noEmit": true,
    "noImplicitAny": true,
    "target": "es2019",
    "types": Array [
      "chrome",
      "jest",
      "offscreencanvas",
      "puppeteer-core",
    ],
  },
  "files": Array [
    "./coverage.js",
    "./environment.js",
    "./globals.d.ts",
    "./jest.config.js",
    "./paths.js",
    "./server.js",
    "./dynamic/inline-override.tests.ts",
    "./dynamic/link-override.tests.ts",
    "./dynamic/style-override.tests.ts",
    "./e2e/devtools.tests.ts",
    "./e2e/toggle.tests.ts",
    "./settings/migration.tests.ts",
    "./settings/sitelist.tests.ts",
  ],
}
`;

exports[`TypeScript project config file should parse and resolve correctly: tests/inject 1`] = `
Object {
  "compileOnSave": false,
  "compilerOptions": Object {
    "allowJs": true,
    "downlevelIteration": true,
    "esModuleInterop": true,
    "inlineSources": true,
    "jsx": "react",
    "jsxFactory": "m",
    "lib": Array [
      "es6",
      "es2017",
      "dom",
      "dom.iterable",
    ],
    "module": "es6",
    "moduleResolution": "node",
    "noEmit": true,
    "noEmitOnError": true,
    "noImplicitAny": true,
    "removeComments": false,
    "sourceMap": true,
    "target": "es2019",
    "types": Array [
      "chrome",
      "offscreencanvas",
      "jasmine",
    ],
  },
  "exclude": Array [
    "./coverage",
  ],
  "files": Array [
    "./karma.conf.ci.js",
    "./karma.conf.js",
    "./types.d.ts",
    "./dynamic/color.tests.ts",
    "./dynamic/fixes.tests.ts",
    "./dynamic/image-analysis.tests.ts",
    "./dynamic/inline-override.tests.ts",
    "./dynamic/link-override.tests.ts",
    "./dynamic/media-query.tests.ts",
    "./dynamic/shadow-dom.tests.ts",
    "./dynamic/style-override.tests.ts",
    "./dynamic/variables.tests.ts",
    "./support/background-stub.ts",
    "./support/customize.ts",
    "./support/polyfills.ts",
    "./support/test-utils.ts",
    "./types/karma-rollup-preprocessor.d.ts",
    "./types/karma.d.ts",
    "./types/rollup-plugin-istanbul2.d.ts",
    "./utils/url.tests.ts",
  ],
}
`;

exports[`TypeScript project config file should parse and resolve correctly: tests/unit 1`] = `
Object {
  "compileOnSave": false,
  "compilerOptions": Object {
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "inlineSourceMap": true,
    "lib": Array [
      "esnext",
      "dom",
    ],
    "module": "commonjs",
    "noImplicitAny": true,
    "skipLibCheck": true,
    "target": "esnext",
    "types": Array [
      "node",
      "jest",
      "jest-extended",
    ],
  },
  "files": Array [
    "./config/config.tests.ts",
    "./config/locales.tests.ts",
    "./generators/utils/parse.tests.ts",
    "./utils/color.tests.ts",
    "./utils/math.tests.ts",
    "./utils/parsing.tests.ts",
    "./utils/promise-barrier.tests.ts",
    "./utils/text.tests.ts",
    "./utils/time.tests.ts",
    "./utils/uid.tests.ts",
  ],
}
`;
